### PR TITLE
Client evaluation of stored procedures with composition

### PR DESCRIPTION
**Pull request by mikary**
@anpete @divega

### CODE REVIEW

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs [hunk 1]
```diff
@@ -104,6 +104,26 @@ protected override Expression VisitEntityQueryable([NotNull] Type elementType)
                         alias,
                         _querySource));
 
+            var composed = true;
+            if (fromSqlAnnotation != null)
+            {
+                if (fromSqlAnnotation.Sql.IndexOf("EXEC", StringComparison.OrdinalIgnoreCase) >= 0)
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: StartsWith?

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs [hunk 2]
```diff
@@ -104,6 +104,26 @@ protected override Expression VisitEntityQueryable([NotNull] Type elementType)
...
+                {
+                    if (QueryModelVisitor.QueryCompilationContext.QueryAnnotations.OfType<IncludeQueryAnnotation>().Any())
+                    {
+                        throw new InvalidOperationException(Strings.StoredProcedureIncludeNotSupported);
+                    }
+                    QueryModelVisitor.RequiresClientFilter = true;
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Can you add a note here that we need to set any other client eval flags that we add.

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs [hunk 3]
```diff
@@ -155,24 +175,15 @@ var keyFactory
...
-                && fromSqlAnnotation.QueryModel.IsIdentityQuery()
-                && !fromSqlAnnotation.QueryModel.ResultOperators.Any())
-            {
-                commandBuilder = new CommandBuilder(new RawSqlQueryGenerator(fromSqlAnnotation.Sql, fromSqlAnnotation.Parameters));
-            }
-            else
-            {
-                commandBuilder = new CommandBuilder(QueryModelVisitor.QueryCompilationContext.CreateSqlQueryGenerator(selectExpression));
-            }
+            var sqlQueryGenerator = (composed)
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Parens

**Comment by anpete** [comment]
:shipit:

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs (hunk 1 above)
Comment (divega) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: What would providers need to do to detect non-composable queries based on a different criteria?

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs (hunk 2 above)
Comment (divega) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Seems like we could use something that expresses than the whole thing cannot be composed over. I am ok with leaving it for a bit later.

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs (hunk 1 above)
Comment (mikary) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Added #2040

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs (hunk 2 above)
Comment (mikary) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Just a side note related to previous discussions. We still need to add the ```SelectExpression``` to the ```QueryModelVisitor``` for ```FromSql``` where the only composition is ```IncludeQueryAnnotation```s.

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs [hunk 4]
```diff
@@ -104,6 +104,27 @@ protected override Expression VisitEntityQueryable([NotNull] Type elementType)
                         alias,
                         _querySource));
 
+            var composed = true;
+            if (fromSqlAnnotation != null)
+            {
+                if (!fromSqlAnnotation.Sql.TrimStart().StartsWith("SELECT", StringComparison.OrdinalIgnoreCase))
```
Comment (divega) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Will this match a stored procedure called SelectProducts? If yes, please file a bug to make it more robust and add negative cases.

Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Regex: ^\s*select\s+

Comment (ErikEJ) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Not good at regex, but check needs to append check for whitespace char following select

Comment (iecompattest) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: Testing

Comment (divega) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: @mikary Did you file a bug on this? Can you help me find it?

Comment (mikary) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs]: @divega I didn't open a new issue for it, but it has been fixed in PR #2053 which is still awaiting review

**Pull request merged**